driver/vhci/vhci_read.c

```diff
@@ -79,30 +82,39 @@ store_urb_reset_dev(PIRP irp, struct urb_req *urbr)
 static NTSTATUS
 store_urb_reset_pipe(PIRP irp, PURB urb, struct urb_req *urbr)
 {
+	/* 1. We clear STALL/HALT feature on endpoint specified by pipe
+	 * 2. We abort/cancel all IRP for given pipe
+	 */
 	struct _URB_PIPE_REQUEST	*urb_rp = &urb->UrbPipeRequest;
 	struct usbip_header	*hdr;
-	int	in, type;
 
 	hdr = get_usbip_hdr_from_read_irp(irp);
 	if (hdr == NULL) {
 		return STATUS_BUFFER_TOO_SMALL;
 	}
 
-	in = PIPE2DIRECT(urb_rp->PipeHandle);
-	type = PIPE2TYPE(urb_rp->PipeHandle);
-	if (type != USB_ENDPOINT_TYPE_BULK && type != USB_ENDPOINT_TYPE_INTERRUPT) {
-		DBGE(DBG_READ, "Error, not a bulk pipe\n");
-		return STATUS_INVALID_PARAMETER;
-	}
+	/*
+	 * Documentation for USB says:
```

Comment:
@koudis :  Check [MSDN comment](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/usb/ns-usb-_urb_header) for URB_FUNCTION_SYNC_RESET_PIPE_AND_CLEAR_STALL
``` 
First, for all pipes except isochronous pipes, this URB sends a CLEAR_FEATURE request to clear the device's ENDPOINT_HALT feature.
```

It seems that URB_FUNCTION_SYNC_RESET_PIPE_AND_CLEAR_STALL is allowed for an iso pipe. 

Type checking code you had removed was a legacy code. Maybe, an old programmer seemed to have a problem that this routine was called on a non-bulk pipe. Also, I do not believe that all drivers or user-level codes behave as usbip-win expects.
I'll revert to a previous revision but improves the code from your work.

Reply:
So for all pipes except iso  URB URB_FUNCTION_SYNC_RESET_PIPE_AND_CLEAR_STALL is valid?
It's possible. I will test it if it works as expected.

### Note
But due to USB doct there are 4 pipe types

- CONTROL
- ISOCHRONOUS
- BULK
- INTERRUPT

For CONTROL is not necessary to clear STALL explicitly.
For ISO URB URB_FUNCTION_SYNC_RESET_PIPE_AND_CLEAR_STALL wil not be send (doue to Msdn doc).

So I think that your assuptions are right. Thanks!

Reply:
@koudis : I think that upper drivers on vhci may call URB_FUNCTION_SYNC_RESET_PIPE_AND_CLEAR_STALL with any type. So the type checking is needed. Please check 35509ef47d56992e4a6ddde178e680b3fc3daa5e .

Reply:
Works after #106